### ISSUE TITLE
Question about sep_loss

### ISSUE THREAD

**Issue by xingyueye** [xiaomi-automl/FairDARTS#7]
@miautoml Sorry to trouble you, I have tried your sep_loss in my darts project. I found that the gap between arch-parameter weights is indeed getting larger. For example, they were originally distributed around 0.1（9 Mixedops）, but now they are basically distributed at the positions of 0.2 and 0.01. However, the problem is that there are usually three or four ops with a weight of 0.2, and the difference is not large. It is also difficult to determine which op to choose.

**Comment by tianbaochou** [comment]
> at there are usually three or four ops with a weight of 0.2, an

Hi, the darts project indeed not fit sep_loss, since the unfair challenge. FairDARTS use this loss because sigmoid function breaks the unfairness.

**Comment by xingyueye** [comment]
Thank you very much. In fact, I independently verified the two points in your paper. From the experiments, the results are great. Next, I will combine the two points to do experiment. In addition, I still have a question. After using sigmood instead of softmax, the update of arch-parameter is slower, so do I need to increase the epoch number of the search?

**Comment by tianbaochou** [comment]
> Thank you very much. In fact, I independently verified the two points in your paper. From the experiments, the results are great. Next, I will combine the two points to do experiment. In addition, I still have a question. After using sigmood instead of softmax, the update of arch-parameter is slower, so do I need to increase the epoch number of the search?

The number of epoch your search also a hyper-parameters. Different ML tasks even different train datasets (for search) maybe use not same epoch.  In our experiments, the CIFAR-10 dataset can reduce epoch  from 50 to 40 or lower. For fair comparation, we use the same epoch as DARTS

**Comment by xingyueye** [comment]
@tianbaochou @miautoml Unintentionally offensive, but I do have a little doubt whether "sigmoid instead of softmax" really breaks the unfairness of Darts. We know that softmax is conducive to speeding up optimization. After using sigmoid instead, it is likely to only slow down the update speed of the model structure. In fact, my experimental results show that the structure update is really very slow after using sigmoid. At the same time, after I increased the arch_lr and arch-gamma parameters in Architect, I can see that the structure update has accelerated, but the number of weights-free op is still getting more and more. In this way, the unfairness of Darts has not been resolved, it is just that after the model update becomes slow, the collapse phenomenon is not displayed under the same search_epoch

**Comment by miautoml** [comment]
Thanks for the feedback. We suggest reading the updated paper which is more clear. As per breaking the unfairness,  there might be some misunderstanding. Using sigmoid, more weight-free ops still appear (also shown in the paper) but they don't cause the same problem since they can co-exist with other ops. We don't mean to reduce the numbers of them, hope this helps.

**Comment by xingyueye** [comment]
@miautoml thanks a lot for your reply, I will continue to follow your follow-up articles and research progress. I agree with your opinion that weight-free ops should co-exit with other ops. But weight-free ops still appear too much in my task, even searched with sep_loss and sigmoid. By the way, I find that adding warmup to fairarts is very effective, and so I got better results. Actually,  warmup is a more common and explainable method to deal with the unfairness of weight-free and other ops.

**Comment by miautoml** [comment]
Yup, early-stage warmup can be useful, but skip connections show their unfair advantages after the **boundary point**.  It would be interesting to check how much importance does warmup have. We would still like to emphasize, as you put

> weight-free ops still appear too much in my task

is that **much** is no longer a problem, it is allowed as they don't cause performance collapse in FairDARTS. However, it is not the case in DARTS.

**Comment by xingyueye** [comment]
@miautoml yes, skip_connections is a really big problem after warmup...lol
As you said:
> is no longer a problem in FairDARTS
How do you prove this point, it‘s not sufficient if only stand on the final accuararcy. Maybe we need a set of experiment like Figure2 in Darts+.

**Comment by miautoml** [comment]
@xingyueye Our final accuracies are averaged on models of 7 independent runs (97.46±0.05). We believe it is adequate to show it doesn't suffer from performance collapse.

**Comment by miautoml** [comment]
@xingyueye As of Fig 2 in DARTS+, I personally don't recommend replicating the issue in the FairDARTS context. Apart from the reason I stated, you ought to fully train 50 models (costly to some extent) at least for CIFAR-10. A better way is to check whether many skip connections come alone (only skip has high `sigmoid(alpha)` ) or co-appear with other ops in the final search epoch. The former is the cause of the collapse.

**Comment by xingyueye** [comment]
@miautoml，well,  I agree with you since you could get such a result (97.46 ± 0.05 for many times test independently), it is enough to prove the stability of search. 
The reproducibility of many Darts' improvement work is too low(!0!), making it hard to verify its effectiveness. But your FairDarts did help to my work, thank you very much

**Comment by miautoml** [comment]
> @miautoml，well, I agree with you since you could get such a result (97.46 ± 0.05 for many times test independently), it is enough to prove the stability of search.
> The reproducibility of many Darts' improvement work is too low(!0!), making it hard to verify its effectiveness. But your FairDarts did help to my work, thank you very much

We are glad it helped. Any other comments, please don't hesitate to share it with us.
